src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs

```diff
@@ -90,26 +94,49 @@ public void Abort()
                 AddWebSocketHeaders(request, secKeyAndSecWebSocketAccept.Key, options);
 
                 // Create the handler for this request and populate it with all of the options.
-                handler = new SocketsHttpHandler();
-                handler.Credentials = options.Credentials;
-                handler.Proxy = options.Proxy;
-                handler.CookieContainer = options.Cookies;
-                handler.SslOptions.RemoteCertificateValidationCallback = options.RemoteCertificateValidationCallback;
-                if (options._clientCertificates?.Count > 0) // use field to avoid lazily initializing the collection
+                // Try to use a shared handler rather than creating a new one just for this request, if
+                // the options are compatible.
+                if (options.Credentials == null &&
+                    options.Proxy == null &&
+                    options.Cookies == null &&
+                    options.RemoteCertificateValidationCallback == null &&
+                    options._clientCertificates?.Count == 0)
+                {
+                    disposeHandler = false;
+                    handler = s_defaultHandler;
+                    if (handler == null)
+                    {
+                        handler = new SocketsHttpHandler() { PooledConnectionLifetime = TimeSpan.Zero };
+                        if (Interlocked.CompareExchange(ref s_defaultHandler, handler, null) != null)
+                        {
+                            handler.Dispose();
+                            handler = s_defaultHandler;
+                        }
+                    }
+                }
+                else
                 {
-                    if (handler.SslOptions.ClientCertificates == null)
+                    handler = new SocketsHttpHandler() { PooledConnectionLifetime = TimeSpan.Zero };
+                    handler.Credentials = options.Credentials;
+                    handler.Proxy = options.Proxy;
+                    handler.CookieContainer = options.Cookies;
+                    handler.SslOptions.RemoteCertificateValidationCallback = options.RemoteCertificateValidationCallback;
+                    if (options._clientCertificates?.Count > 0) // use field to avoid lazily initializing the collection
                     {
-                        handler.SslOptions.ClientCertificates = new X509Certificate2Collection();
+                        if (handler.SslOptions.ClientCertificates == null)
```

Comment:
Not that it matters and it is perhaps good to protect against possible future changes, but isn't this always going to be true?

Reply:
> but isn't this always going to be true?

Currently, yes.  I went back and forth on whether to just assert this or not.  But since it's not in the same assembly, I thought it best to check.  I could be convinced to change it, though.

Reply:
Given that for it to be false the constructor needs to be making one extra allocation, that likely will be useless, I doubt that we will ever do such change :smile:. That said I'm fine as it is.

Reply:
Sometimes an assert is useful simply to document that the reader can make an assumption.

Reply:
My latest commit has it using an assert.